// @tommyvez/passfort/dist/cli.js
#!/usr/bin/env node
import { readFileSync, writeFileSync, existsSync } from 'fs';
import { resolve, join } from 'path';

// src/crypto.ts
var PBKDF2_ITERATIONS = 1e5;
var SALT_LENGTH = 16;
var HASH_LENGTH = 32;
function toBase64Url(bytes) {
  const base64 = btoa(String.fromCharCode(...bytes));
  return base64.replace(/\+/g, "-").replace(/\//g, "_").replace(/=+$/, "");
}
async function hashPassword(password) {
  const salt = crypto.getRandomValues(new Uint8Array(SALT_LENGTH));
  const encoder = new TextEncoder();
  const passwordKey = await crypto.subtle.importKey(
    "raw",
    encoder.encode(password),
    { name: "PBKDF2" },
    false,
    ["deriveBits"]
  );
  const hash = await crypto.subtle.deriveBits(
    {
      name: "PBKDF2",
      hash: "SHA-256",
      salt,
      iterations: PBKDF2_ITERATIONS
    },
    passwordKey,
    HASH_LENGTH * 8
  );
  const hashBytes = new Uint8Array(hash);
  return `pbkdf2:${PBKDF2_ITERATIONS}:${toBase64Url(salt)}:${toBase64Url(hashBytes)}`;
}
var PACKAGE_IMPORT = "@tommyvez/passfort/next";
var MATCHER_SNIPPET = `import { withPasswordProtect } from '${PACKAGE_IMPORT}';

export default withPasswordProtect({ protectAll: true });

export const config = {
  matcher: ['/((?!api|_next|favicon.ico).*)'],
};
`;
var MATCHER_BLOCK_SNIPPET = `import { withPasswordProtect } from '${PACKAGE_IMPORT}';

export default withPasswordProtect({ protectAll: true, blockOnly: true });

export const config = {
  matcher: ['/((?!api|_next|favicon.ico).*)'],
};
`;
var PROXY_SNIPPET = `import { withPasswordProtect } from '${PACKAGE_IMPORT}';

export const proxy = withPasswordProtect({ protectAll: true });

export const config = {
  matcher: ['/((?!api|_next|favicon.ico).*)'],
};
`;
var PROXY_BLOCK_SNIPPET = `import { withPasswordProtect } from '${PACKAGE_IMPORT}';

export const proxy = withPasswordProtect({ protectAll: true, blockOnly: true });

export const config = {
  matcher: ['/((?!api|_next|favicon.ico).*)'],
};
`;
function findNextProjectRoot(startDir) {
  let dir = resolve(startDir);
  for (let i = 0; i < 20; i++) {
    const pkgPath = join(dir, "package.json");
    if (existsSync(pkgPath)) {
      try {
        const pkg = JSON.parse(readFileSync(pkgPath, "utf-8"));
        const deps = { ...pkg.dependencies, ...pkg.devDependencies };
        if (deps && (deps.next || deps["next"])) return dir;
      } catch {
      }
    }
    const parent = resolve(dir, "..");
    if (parent === dir) return null;
    dir = parent;
  }
  return null;
}
function findMiddlewarePath(root) {
  const atRoot = join(root, "middleware.ts");
  if (existsSync(atRoot)) return atRoot;
  if (existsSync(join(root, "middleware.js")))
    return join(root, "middleware.js");
  const atSrc = join(root, "src", "middleware.ts");
  if (existsSync(atSrc)) return atSrc;
  if (existsSync(join(root, "src", "middleware.js")))
    return join(root, "src", "middleware.js");
  return null;
}
function getMiddlewareWritePath(root) {
  const hasSrcApp = existsSync(join(root, "src", "app")) || existsSync(join(root, "src", "pages"));
  return hasSrcApp ? join(root, "src", "middleware.ts") : join(root, "middleware.ts");
}
function generateMiddlewareContent(opts) {
  if (opts.block) return MATCHER_BLOCK_SNIPPET.trim();
  if (opts.paths && opts.paths.length > 0) {
    const pathList = opts.paths.map((p) => `'${p.startsWith("/") ? p : "/" + p}'`).join(", ");
    const matcherList = opts.paths.map((p) => `'${p.startsWith("/") ? p : "/" + p}/:path*'`).join(", ");
    return `import { withPasswordProtect } from '${PACKAGE_IMPORT}';

export default withPasswordProtect({
  paths: [${pathList}],
});

export const config = {
  matcher: [${matcherList}],
};
`;
  }
  return MATCHER_SNIPPET.trim();
}
function generateProxyContent(opts) {
  if (opts.block) return PROXY_BLOCK_SNIPPET.trim();
  if (opts.paths && opts.paths.length > 0) {
    const pathList = opts.paths.map((p) => `'${p.startsWith("/") ? p : "/" + p}'`).join(", ");
    const matcherList = opts.paths.map((p) => `'${p.startsWith("/") ? p : "/" + p}/:path*'`).join(", ");
    return `import { withPasswordProtect } from '${PACKAGE_IMPORT}';

export const proxy = withPasswordProtect({
  paths: [${pathList}],
});

export const config = {
  matcher: [${matcherList}],
};
`;
  }
  return PROXY_SNIPPET.trim();
}
function findProxyPath(root) {
  const atRoot = join(root, "proxy.ts");
  if (existsSync(atRoot)) return atRoot;
  if (existsSync(join(root, "proxy.js"))) return join(root, "proxy.js");
  const atSrc = join(root, "src", "proxy.ts");
  if (existsSync(atSrc)) return atSrc;
  if (existsSync(join(root, "src", "proxy.js")))
    return join(root, "src", "proxy.js");
  return null;
}
function getProxyWritePath(root) {
  const hasSrcApp = existsSync(join(root, "src", "app")) || existsSync(join(root, "src", "pages"));
  return hasSrcApp ? join(root, "src", "proxy.ts") : join(root, "proxy.ts");
}
function contentHasPassfort(content) {
  return content.includes("withPasswordProtect") || content.includes("@tommyvez/passfort/next") || content.includes("passfort/next");
}
function runInit(options) {
  const { startDir, block, paths, useProxy } = options;
  const root = findNextProjectRoot(startDir);
  if (!root) return { ok: false, reason: "not_next" };
  if (useProxy) {
    const existingPath2 = findProxyPath(root);
    if (existingPath2) {
      const content3 = readFileSync(existingPath2, "utf-8");
      if (contentHasPassfort(content3)) {
        return { ok: false, reason: "already_set_up", path: existingPath2 };
      }
      return { ok: false, reason: "proxy_exists", path: existingPath2 };
    }
    const outPath2 = getProxyWritePath(root);
    const content2 = generateProxyContent({ block, paths });
    writeFileSync(outPath2, content2 + "\n", "utf-8");
    return { ok: true, path: outPath2 };
  }
  const existingPath = findMiddlewarePath(root);
  if (existingPath) {
    const content2 = readFileSync(existingPath, "utf-8");
    if (contentHasPassfort(content2)) {
      return { ok: false, reason: "already_set_up", path: existingPath };
    }
    return { ok: false, reason: "middleware_exists", path: existingPath };
  }
  const outPath = getMiddlewareWritePath(root);
  const content = generateMiddlewareContent({ block, paths });
  writeFileSync(outPath, content + "\n", "utf-8");
  return { ok: true, path: outPath };
}

// src/cli.ts
var args = process.argv.slice(2);
var cmd = args[0];
if (cmd === "hash") {
  const password = args[1];
  if (!password) {
    console.error('Usage: passfort hash "your-password"');
    process.exit(1);
  }
  const hash = await hashPassword(password);
  console.log("\nAdd to your Vercel environment variables:\n");
  console.log(`PASSFORT_HASH=${hash}`);
  console.log("(or PASSWORD_PROTECT_HASH for backward compatibility)\n");
  console.log("Also set PASSFORT_SECRET (min 16 chars):");
  console.log("  openssl rand -base64 24\n");
  process.exit(0);
}
if (cmd === "matcher") {
  const block = args.includes("--block") || args.includes("-b");
  const useProxy = args.includes("--proxy") || args.includes("-p");
  const fileHint = useProxy ? "proxy.ts (Next 16+)" : "middleware.ts";
  if (block) {
    console.log(
      `
# Block entire site (maintenance mode). Paste into ${fileHint}
`
    );
    console.log(useProxy ? PROXY_BLOCK_SNIPPET : MATCHER_BLOCK_SNIPPET);
    console.log("");
  } else {
    console.log(`
# Protect entire site. Paste into ${fileHint}
`);
    console.log(useProxy ? PROXY_SNIPPET : MATCHER_SNIPPET);
    console.log(
      "\nThen set PASSFORT_PASSWORD and PASSFORT_SECRET in Vercel.\n"
    );
  }
  process.exit(0);
}
if (cmd === "init") {
  if (args.includes("--help") || args.includes("-h")) {
    console.log(`
passfort init - Create middleware.ts (or proxy.ts) and add passfort

  npx passfort init              Protect entire site (middleware)
  npx passfort init --proxy      Next 16+: create proxy.ts
  npx passfort init --paths=/admin,/dashboard
  npx passfort init --block      Maintenance mode (503)

Options:
  --paths=/a,/b   Protect only these paths
  --block         Block entire site with 503 (no form)
  --proxy         Create proxy.ts for Next 16+ (named export proxy)
`);
    process.exit(0);
  }
  const cwd = process.cwd();
  const block = args.includes("--block") || args.includes("-b");
  const useProxy = args.includes("--proxy") || args.includes("-p");
  const pathsArg = args.find((a) => a.startsWith("--paths="));
  const paths = pathsArg ? pathsArg.slice("--paths=".length).split(",").map((p) => p.trim()).filter(Boolean) : void 0;
  const result = runInit({ startDir: cwd, block, paths, useProxy });
  if (result.ok) {
    console.log("Created " + result.path);
    if (block) {
      console.log(
        "Block-only mode (503). Set PASSFORT_BLOCK_ONLY=true in production if you use env-based config."
      );
    } else {
      console.log(
        "Set PASSFORT_PASSWORD and PASSFORT_SECRET in your environment (e.g. Vercel)."
      );
    }
    process.exit(0);
  }
  if (result.reason === "not_next") {
    console.error(
      "Not a Next.js project (no package.json with next dependency found from " + cwd
    );
    process.exit(1);
  }
  if (result.reason === "already_set_up") {
    console.log("passfort is already set up in " + result.path);
    process.exit(0);
  }
  if (result.reason === "middleware_exists") {
    console.log("middleware already exists at " + result.path);
    console.log(
      "Use 'npx passfort matcher' to see the snippet and merge manually, or back up the file and run 'npx passfort init' again."
    );
    process.exit(1);
  }
  if (result.reason === "proxy_exists") {
    console.log("proxy already exists at " + result.path);
    console.log(
      "Use 'npx passfort matcher --proxy' to see the snippet and merge manually, or back up the file and run 'npx passfort init --proxy' again."
    );
    process.exit(1);
  }
  process.exit(1);
}
if (cmd === "--help" || cmd === "-h" || !cmd) {
  console.log(`
passfort - Password protection for Vercel

Commands:
  init [options]        Create middleware.ts or proxy.ts and add passfort
  hash <password>       Generate PBKDF2 hash for production use
  matcher [options]     Print snippet for middleware.ts or proxy.ts

Init options:
  --paths=/admin,/foo   Protect specific paths only (default: entire site)
  --block               Block entire site with 503 (maintenance mode)
  --proxy               Next 16+: create proxy.ts (named export proxy)

Matcher options:
  --block               Block-only (503) snippet
  --proxy               Proxy snippet for proxy.ts (Next 16+)

Quick start:
  1. pnpm add @tommyvez/passfort
  2. npx passfort init              # create middleware (entire site protected)
     or npx passfort init --proxy   # Next 16+: create proxy.ts
  3. Set env: PASSFORT_PASSWORD, PASSFORT_SECRET

Manual: npx passfort matcher        # paste into middleware.ts
        npx passfort matcher --proxy   # paste into proxy.ts (Next 16+)

Production:
  npx passfort hash "mypassword"
  Set PASSFORT_HASH and PASSFORT_SECRET in Vercel
`);
  process.exit(0);
}
console.error(`Unknown command: ${cmd}`);
process.exit(1);
