// @tommyvez/passfort/dist/index.js
// src/config.ts
var DEFAULT_SESSION_DURATION = 60 * 60 * 24 * 7;
var MIN_SECRET_LENGTH = 16;
function loadConfig(env = process.env) {
  const enabled = env.PASSFORT_ENABLED ?? env.PASSWORD_PROTECT_ENABLED ?? env.VERCEL_PASSWORD_ENABLED;
  if (enabled === "false" || enabled === "0") {
    return null;
  }
  const blockOnly = env.PASSFORT_BLOCK_ONLY === "true" || env.PASSFORT_BLOCK_ONLY === "1" || env.PASSWORD_PROTECT_BLOCK_ONLY === "true" || env.PASSWORD_PROTECT_BLOCK_ONLY === "1" || env.VERCEL_PASSWORD_BLOCK_ONLY === "true" || env.VERCEL_PASSWORD_BLOCK_ONLY === "1";
  if (blockOnly) {
    return {
      blockOnly: true,
      secret: void 0,
      sessionDuration: DEFAULT_SESSION_DURATION
    };
  }
  const secret = env.PASSFORT_SECRET ?? env.PASSWORD_PROTECT_SECRET ?? env.VERCEL_PASSWORD_SECRET;
  if (!secret || secret.length < MIN_SECRET_LENGTH) {
    return null;
  }
  const password = env.PASSFORT_PASSWORD ?? env.PASSWORD_PROTECT_PASSWORD ?? env.VERCEL_PASSWORD;
  const hash = env.PASSFORT_HASH ?? env.PASSWORD_PROTECT_HASH ?? env.VERCEL_PASSWORD_HASH;
  if (!password && !hash) {
    return null;
  }
  const durationStr = env.PASSFORT_SESSION_DURATION ?? env.PASSWORD_PROTECT_SESSION_DURATION ?? env.VERCEL_PASSWORD_SESSION_DURATION;
  const sessionDuration = durationStr ? parseInt(durationStr, 10) : DEFAULT_SESSION_DURATION;
  const mode = env.PASSFORT_MODE ?? env.PASSWORD_PROTECT_MODE ?? env.VERCEL_PASSWORD_MODE ?? "form";
  const form = {};
  const formTitle = env.PASSFORT_FORM_TITLE ?? env.PASSWORD_PROTECT_FORM_TITLE;
  const formDesc = env.PASSFORT_FORM_DESCRIPTION ?? env.PASSWORD_PROTECT_FORM_DESCRIPTION;
  const formPlaceholder = env.PASSFORT_FORM_PLACEHOLDER ?? env.PASSWORD_PROTECT_FORM_PLACEHOLDER;
  const formButton = env.PASSFORT_FORM_BUTTON ?? env.PASSWORD_PROTECT_FORM_BUTTON;
  const formTheme = env.PASSFORT_FORM_THEME ?? env.PASSWORD_PROTECT_FORM_THEME;
  if (formTitle) form.title = formTitle;
  if (formDesc) form.description = formDesc;
  if (formPlaceholder) form.placeholder = formPlaceholder;
  if (formButton) form.buttonText = formButton;
  if (formTheme === "light") form.theme = "light";
  const loginPath = env.PASSFORT_LOGIN_PATH ?? env.PASSWORD_PROTECT_LOGIN_PATH ?? env.VERCEL_PASSWORD_LOGIN_PATH;
  const protectAll = env.PASSFORT_ALL === "true" || env.PASSFORT_ALL === "1" || env.PASSWORD_PROTECT_ALL === "true" || env.PASSWORD_PROTECT_ALL === "1" || env.VERCEL_PASSWORD_PROTECT_ALL === "true" || env.VERCEL_PASSWORD_PROTECT_ALL === "1";
  const excludePathsStr = env.PASSFORT_EXCLUDE_PATHS ?? env.PASSWORD_PROTECT_EXCLUDE_PATHS ?? env.VERCEL_PASSWORD_EXCLUDE_PATHS;
  const excludePaths = excludePathsStr ? excludePathsStr.split(",").map((p) => p.trim()).filter(Boolean) : void 0;
  const rateLimitMaxStr = env.PASSFORT_RATE_LIMIT_MAX ?? env.PASSWORD_PROTECT_RATE_LIMIT_MAX;
  const rateLimitWindowStr = env.PASSFORT_RATE_LIMIT_WINDOW_MS ?? env.PASSWORD_PROTECT_RATE_LIMIT_WINDOW_MS;
  const rateLimitMax = rateLimitMaxStr ? parseInt(rateLimitMaxStr, 10) : 10;
  const rateLimitWindowMs = rateLimitWindowStr ? parseInt(rateLimitWindowStr, 10) : 6e4;
  return {
    password: password ?? void 0,
    hash: hash ?? void 0,
    secret: secret ?? void 0,
    sessionDuration: Number.isFinite(sessionDuration) ? sessionDuration : DEFAULT_SESSION_DURATION,
    mode: mode === "basic" ? "basic" : "form",
    form: Object.keys(form).length > 0 ? form : void 0,
    loginPath: loginPath ?? void 0,
    protectAll: protectAll || void 0,
    excludePaths,
    rateLimitMax: Number.isFinite(rateLimitMax) && rateLimitMax >= 0 ? rateLimitMax : 10,
    rateLimitWindowMs: Number.isFinite(rateLimitWindowMs) && rateLimitWindowMs > 0 ? rateLimitWindowMs : 6e4
  };
}

// src/crypto.ts
var PBKDF2_ITERATIONS = 1e5;
var SALT_LENGTH = 16;
var HASH_LENGTH = 32;
function timingSafeEqual(a, b) {
  if (typeof a !== "string" || typeof b !== "string") {
    return false;
  }
  if (a.length !== b.length) {
    return false;
  }
  let result = 0;
  for (let i = 0; i < a.length; i++) {
    result |= a.charCodeAt(i) ^ b.charCodeAt(i);
  }
  return result === 0;
}
function toBase64Url(bytes) {
  const base64 = btoa(String.fromCharCode(...bytes));
  return base64.replace(/\+/g, "-").replace(/\//g, "_").replace(/=+$/, "");
}
function fromBase64Url(str) {
  let base64 = str.replace(/-/g, "+").replace(/_/g, "/");
  const padding = (4 - base64.length % 4) % 4;
  base64 += "=".repeat(padding);
  const binary = atob(base64);
  return new Uint8Array([...binary].map((c) => c.charCodeAt(0)));
}
async function hashPassword(password) {
  const salt = crypto.getRandomValues(new Uint8Array(SALT_LENGTH));
  const encoder = new TextEncoder();
  const passwordKey = await crypto.subtle.importKey(
    "raw",
    encoder.encode(password),
    { name: "PBKDF2" },
    false,
    ["deriveBits"]
  );
  const hash = await crypto.subtle.deriveBits(
    {
      name: "PBKDF2",
      hash: "SHA-256",
      salt,
      iterations: PBKDF2_ITERATIONS
    },
    passwordKey,
    HASH_LENGTH * 8
  );
  const hashBytes = new Uint8Array(hash);
  return `pbkdf2:${PBKDF2_ITERATIONS}:${toBase64Url(salt)}:${toBase64Url(hashBytes)}`;
}
async function verifyPasswordHash(password, hashString) {
  try {
    const parts = hashString.split(":");
    if (parts.length !== 4 || parts[0] !== "pbkdf2") {
      return false;
    }
    const iterations = parseInt(parts[1], 10);
    if (!Number.isFinite(iterations) || iterations < 1e4) {
      return false;
    }
    const salt = fromBase64Url(parts[2]);
    const expectedHash = fromBase64Url(parts[3]);
    if (salt.length < 8 || expectedHash.length !== HASH_LENGTH) {
      return false;
    }
    const encoder = new TextEncoder();
    const passwordKey = await crypto.subtle.importKey(
      "raw",
      encoder.encode(password),
      { name: "PBKDF2" },
      false,
      ["deriveBits"]
    );
    const derived = await crypto.subtle.deriveBits(
      {
        name: "PBKDF2",
        hash: "SHA-256",
        salt,
        iterations
      },
      passwordKey,
      HASH_LENGTH * 8
    );
    const derivedBytes = new Uint8Array(derived);
    if (derivedBytes.length !== expectedHash.length) return false;
    let result = 0;
    for (let i = 0; i < derivedBytes.length; i++) {
      result |= derivedBytes[i] ^ expectedHash[i];
    }
    return result === 0;
  } catch {
    return false;
  }
}
async function signPayload(secret, payload) {
  const encoder = new TextEncoder();
  const key = await crypto.subtle.importKey(
    "raw",
    encoder.encode(secret),
    { name: "HMAC", hash: "SHA-256" },
    false,
    ["sign"]
  );
  const signature = await crypto.subtle.sign(
    "HMAC",
    key,
    encoder.encode(payload)
  );
  return toBase64Url(new Uint8Array(signature));
}
async function verifySignature(secret, payload, signature) {
  const expected = await signPayload(secret, payload);
  return timingSafeEqual(signature, expected);
}

// src/auth.ts
async function verifyPassword(submitted, config) {
  if (!submitted || submitted.length > 256) {
    return false;
  }
  if (config.hash) {
    return verifyPasswordHash(submitted, config.hash);
  }
  if (config.password) {
    return timingSafeEqual(submitted, config.password);
  }
  return false;
}
async function createSession(secret, durationSeconds) {
  const expires = Math.floor(Date.now() / 1e3) + durationSeconds;
  const payload = `e=${expires}`;
  const signature = await signPayload(secret, payload);
  return { payload, signature };
}
async function validateSession(cookieValue, secret) {
  if (!cookieValue || !secret) return false;
  const [payload, signature] = cookieValue.split(".");
  if (!payload || !signature) return false;
  const valid = await verifySignature(secret, payload, signature);
  if (!valid) return false;
  const match = payload.match(/e=(\d+)/);
  if (!match) return false;
  const expires = parseInt(match[1], 10);
  return Date.now() / 1e3 < expires;
}

// src/password-form.html.ts
var DEFAULTS = {
  title: "Password Required",
  description: "This page is protected. Enter the password to continue.",
  placeholder: "Password",
  buttonText: "Continue",
  theme: "dark"
};
var THEMES = {
  dark: {
    body: "#18181b",
    bodyColor: "#e4e4e7",
    card: "#27272a",
    heading: "#ffffff",
    muted: "#a1a1aa",
    inputBg: "#18181b",
    inputBorder: "#52525b",
    inputColor: "#e4e4e7",
    focusBorder: "#ea580c",
    focusRing: "rgba(234,88,12,0.25)",
    button: "#ea580c",
    buttonHover: "#c2410c",
    error: "#ef4444"
  },
  light: {
    body: "#fafafa",
    bodyColor: "#18181b",
    card: "#ffffff",
    heading: "#18181b",
    muted: "#71717a",
    inputBg: "#ffffff",
    inputBorder: "#e4e4e7",
    inputColor: "#18181b",
    focusBorder: "#ea580c",
    focusRing: "rgba(234,88,12,0.2)",
    button: "#ea580c",
    buttonHover: "#c2410c",
    error: "#dc2626"
  }
};
function getPasswordFormHtml(options = {}) {
  const {
    error = false,
    returnUrl = "/",
    title = DEFAULTS.title,
    description = DEFAULTS.description,
    placeholder = DEFAULTS.placeholder,
    buttonText = DEFAULTS.buttonText,
    theme = DEFAULTS.theme
  } = options;
  const colors = THEMES[theme];
  return `<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <title>${escapeHtml(title)}</title>
  <style>
    * { box-sizing: border-box; margin: 0; padding: 0; }
    body {
      font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, sans-serif;
      background: ${colors.body};
      color: ${colors.bodyColor};
      min-height: 100vh;
      display: flex;
      align-items: center;
      justify-content: center;
      padding: 1.5rem;
      -webkit-font-smoothing: antialiased;
      line-height: 1.5;
    }
    .card {
      background: ${colors.card};
      border-radius: 8px;
      padding: 2rem;
      width: 100%;
      max-width: 360px;
      border: 1px solid ${colors.inputBorder};
      box-shadow: 0 1px 2px rgba(0,0,0,0.05);
    }
    h1 {
      font-size: 1.25rem;
      font-weight: 500;
      margin-bottom: 0.5rem;
      color: ${colors.heading};
      letter-spacing: -0.02em;
    }
    p { color: ${colors.muted}; font-size: 0.875rem; margin-bottom: 1.25rem; line-height: 1.5; }
    form { display: flex; flex-direction: column; gap: 1rem; }
    input[type="password"] {
      width: 100%;
      padding: 0.625rem 1rem;
      font-size: 0.9375rem;
      border: 1px solid ${colors.inputBorder};
      border-radius: 6px;
      background: ${colors.inputBg};
      color: ${colors.inputColor};
      transition: border-color 0.15s, box-shadow 0.15s;
    }
    input[type="password"]:focus {
      outline: none;
      border-color: ${colors.focusBorder};
      box-shadow: 0 0 0 2px ${colors.focusRing};
    }
    input[type="password"]::placeholder {
      color: ${colors.muted};
    }
    button {
      padding: 0.625rem 1.25rem;
      font-size: 0.875rem;
      font-weight: 500;
      background: ${colors.button};
      color: white;
      border: none;
      border-radius: 6px;
      cursor: pointer;
      transition: background 0.15s;
    }
    button:hover { background: ${colors.buttonHover}; }
    button:focus {
      outline: none;
      box-shadow: 0 0 0 2px ${colors.focusRing};
    }
    .error {
      color: ${colors.error};
      font-size: 0.8125rem;
      margin-top: -0.25rem;
    }
    .back {
      margin-top: 1.25rem;
      padding-top: 1rem;
      border-top: 1px solid ${colors.inputBorder};
    }
    .back a {
      color: ${colors.muted};
      font-size: 0.875rem;
      text-decoration: none;
      display: inline-block;
      min-height: 2.75rem;
      line-height: 2.75rem;
      padding: 0 0.25rem;
      -webkit-tap-highlight-color: transparent;
    }
    .back a:hover { color: ${colors.bodyColor}; }
    .back a:focus {
      outline: none;
      color: ${colors.focusBorder};
    }
  </style>
</head>
<body>
  <div class="card">
    <h1>${escapeHtml(title)}</h1>
    <p>${escapeHtml(description)}</p>
    <form method="POST" action="">
      <input type="hidden" name="return_url" value="${escapeHtml(returnUrl)}">
      <input type="password" name="password" placeholder="${escapeHtml(placeholder)}" required autocomplete="current-password" autofocus>
      ${error ? `<p class="error">Incorrect password. Try again.</p>` : ""}
      <button type="submit">${escapeHtml(buttonText)}</button>
    </form>
    <p class="back"><a href="/">\u2190 Back to home</a></p>
  </div>
</body>
</html>`;
}
function escapeHtml(str) {
  return str.replace(/&/g, "&amp;").replace(/</g, "&lt;").replace(/>/g, "&gt;").replace(/"/g, "&quot;").replace(/'/g, "&#39;");
}

// src/rate-limit.ts
var store = /* @__PURE__ */ new Map();
var PRUNE_AFTER_MS = 12e4;
var lastPrune = 0;
function prune() {
  const now = Date.now();
  if (now - lastPrune < PRUNE_AFTER_MS) return;
  lastPrune = now;
  for (const [key, entry] of store.entries()) {
    if (entry.resetAt < now) store.delete(key);
  }
}
function checkRateLimit(key, maxAttempts, windowMs) {
  if (maxAttempts <= 0 || windowMs <= 0) return null;
  prune();
  const now = Date.now();
  const entry = store.get(key);
  if (!entry) {
    store.set(key, { count: 1, resetAt: now + windowMs });
    return null;
  }
  if (entry.resetAt < now) {
    store.set(key, { count: 1, resetAt: now + windowMs });
    return null;
  }
  entry.count += 1;
  if (entry.count > maxAttempts) {
    const retryAfterSec = Math.ceil((entry.resetAt - now) / 1e3);
    return Math.max(1, retryAfterSec);
  }
  return null;
}

// src/handler.ts
var COOKIE_NAME = "__password_protect";
var COOKIE_MAX_AGE = 60 * 60 * 24 * 7;
async function handlePasswordProtect(options) {
  const { request, config: configOverride, env } = options;
  const baseConfig = loadConfig(env);
  const config = baseConfig ? { ...baseConfig, ...configOverride } : configOverride;
  if (!config) {
    return null;
  }
  if (config.blockOnly) {
    return new Response(
      `<!DOCTYPE html><html><head><meta charset="utf-8"><title>Unavailable</title></head><body style="font-family:system-ui,sans-serif;display:flex;align-items:center;justify-content:center;min-height:100vh;margin:0"><p style="color:#5f6368">Temporarily unavailable.</p></body></html>`,
      {
        status: 503,
        headers: {
          "Content-Type": "text/html; charset=utf-8",
          "Cache-Control": "no-store",
          "Retry-After": "300"
        }
      }
    );
  }
  if (!config.secret) {
    return null;
  }
  const url = new URL(request.url);
  if (request.method === "GET" && url.searchParams.get("passfort_logout") === "1") {
    const clearCookie = `${COOKIE_NAME}=; Path=/; Max-Age=0; HttpOnly; SameSite=Lax${url.protocol === "https:" ? "; Secure" : ""}`;
    return new Response(null, {
      status: 302,
      headers: {
        Location: `${url.origin}/`,
        "Set-Cookie": clearCookie,
        "Cache-Control": "no-store"
      }
    });
  }
  const cookieHeader = request.headers.get("cookie") ?? "";
  const sessionCookie = parseCookie(cookieHeader, COOKIE_NAME);
  if (sessionCookie) {
    const valid = await validateSession(sessionCookie, config.secret);
    if (valid) {
      return null;
    }
  }
  if (config.mode === "basic") {
    return await handleBasicAuth(request, config);
  }
  if (config.loginPath && request.method === "GET") {
    const loginPath = config.loginPath.startsWith("/") ? config.loginPath : `/${config.loginPath}`;
    if (url.pathname === loginPath || url.pathname === `${loginPath}/`) {
      return null;
    }
  }
  if (request.method === "POST") {
    const clientIp = request.headers.get("x-forwarded-for")?.split(",")[0]?.trim() ?? request.headers.get("x-real-ip") ?? "unknown";
    const maxAttempts = config.rateLimitMax ?? 10;
    const windowMs = config.rateLimitWindowMs ?? 6e4;
    const retryAfterSec = checkRateLimit(clientIp, maxAttempts, windowMs);
    if (retryAfterSec !== null) {
      return new Response(null, {
        status: 429,
        headers: {
          "Retry-After": String(retryAfterSec),
          "Cache-Control": "no-store"
        }
      });
    }
    const formData = await request.formData().catch(() => null);
    const password = formData?.get("password");
    const returnUrl = formData?.get("return_url") || url.pathname || "/";
    if (typeof password === "string") {
      const valid = await verifyPassword(password, config);
      if (valid) {
        const { payload, signature } = await createSession(
          config.secret,
          config.sessionDuration ?? COOKIE_MAX_AGE
        );
        const cookieValue = `${payload}.${signature}`;
        const maxAge = config.sessionDuration ?? COOKIE_MAX_AGE;
        const redirectPath = (returnUrl.startsWith("/") ? returnUrl : `/${returnUrl}`) || "/";
        const safePath = redirectPath.startsWith("//") ? "/" : redirectPath;
        const redirectUrl = new URL(safePath, url.origin).toString();
        const isSecure = url.protocol === "https:";
        const cookieFlags = `Path=/; HttpOnly; SameSite=Lax; Max-Age=${maxAge}${isSecure ? "; Secure" : ""}`;
        return new Response(null, {
          status: 302,
          headers: {
            Location: redirectUrl,
            "Set-Cookie": `${COOKIE_NAME}=${cookieValue}; ${cookieFlags}`
          }
        });
      }
    }
  }
  if (config.loginPath) {
    const loginPath = config.loginPath.startsWith("/") ? config.loginPath : `/${config.loginPath}`;
    const returnUrl = encodeURIComponent(url.pathname + url.search);
    return new Response(null, {
      status: 302,
      headers: {
        Location: `${loginPath}?return_url=${returnUrl}`,
        "Cache-Control": "no-store"
      }
    });
  }
  const formOptions = config.form ?? {};
  return new Response(
    getPasswordFormHtml({
      error: request.method === "POST",
      returnUrl: url.pathname + url.search,
      title: formOptions.title,
      description: formOptions.description,
      placeholder: formOptions.placeholder,
      buttonText: formOptions.buttonText,
      theme: formOptions.theme
    }),
    {
      status: 401,
      headers: {
        "Content-Type": "text/html; charset=utf-8",
        "Cache-Control": "no-store"
      }
    }
  );
}
async function handleBasicAuth(request, config) {
  const authHeader = request.headers.get("authorization");
  if (authHeader?.startsWith("Basic ")) {
    try {
      const base64 = authHeader.slice(6);
      const decoded = atob(base64);
      const colonIndex = decoded.indexOf(":");
      if (colonIndex > 0) {
        const password = decoded.slice(colonIndex + 1);
        const valid = await verifyPassword(password, config);
        if (valid) {
          return null;
        }
      }
    } catch {
    }
  }
  return new Response("Authentication required", {
    status: 401,
    headers: {
      "WWW-Authenticate": 'Basic realm="Password Protected"',
      "Cache-Control": "no-store"
    }
  });
}
function parseCookie(cookieHeader, name) {
  const parts = cookieHeader.split(";");
  for (const part of parts) {
    const [key, ...valueParts] = part.trim().split("=");
    if (key === name && valueParts.length > 0) {
      return valueParts.join("=").trim();
    }
  }
  return null;
}

export { handlePasswordProtect, hashPassword, loadConfig, verifyPassword, verifyPasswordHash };
